ocaml/xapi/repository_helpers.ml

```diff
@@ -875,38 +891,44 @@ let remove_repo_conf_file repo_name =
   in
   Unixext.unlink_safe path
 
-let write_yum_config ?(source_url = None) binary_url repo_name =
+let write_yum_config ?(source_url = None) ?(repo_gpgcheck = true) binary_url
+    repo_name gpgkey_name =
   let file_path =
     Filename.concat !Xapi_globs.yum_repos_config_dir (repo_name ^ ".repo")
   in
-  let gpgkey_path =
-    Filename.concat !Xapi_globs.rpm_gpgkey_dir
-      !Xapi_globs.repository_gpgkey_name
+  let opt_repo_gpgcheck, opt_gpgcheck, opt_gpgkey =
+    if !Xapi_globs.repository_gpgcheck then (
+      let gpgkey_path =
+        Filename.concat !Xapi_globs.rpm_gpgkey_dir gpgkey_name
+      in
+      if not (Sys.file_exists gpgkey_path) then
+        raise
+          Api_errors.(
+            Server_error (internal_error, ["gpg key file does not exist"])
+          ) ;
+      if not ((Unix.lstat gpgkey_path).Unix.st_kind = Unix.S_REG) then
+        raise
+          Api_errors.(
+            Server_error (internal_error, ["gpg key file is not a regular file"])
+          ) ;
+      let opt_gpgkey' = Printf.sprintf "gpgkey=file://%s" gpgkey_path in
+      (* Will not check the metadata in local repository on cooridnator *)
+      if repo_gpgcheck then
+        ("repo_gpgcheck=1", "gpgcheck=1", opt_gpgkey')
+      else
+        ("repo_gpgcheck=0", "gpgcheck=1", opt_gpgkey')
```

Comment:
Why would we ever want to not check the GPG key? I think the repository should always be signed with a key, even if just a test key and for testing purposes we could point the host to the right one.
There is a danger here that we might accidentally turn off GPG verification off in production for the sake of testing, I'd rather we always have gpgcheck=1 in the config.
If we really have to then we can add a single boolean in xapi.conf to override this and read it at the point we generate the file instead of plumbing it through lots of function calls with default values

Reply:
For the 'gpgcheck', this was originated from very early stage we were using dummy packages for testing only purpose. And I believe it is being used still.
The current logic is just to be determined by `!Xapi_globs.repository_gpgcheck` which is a config can be configured in xapi.conf.

For the 'repo_gpgcheck', this is used to check meta data of a repo. This works when Toolstack syncs repo from remote repository. After syncing, Toolstack creates a local repository. During this process, the signing is not done for simplicity (otherwise a key has to be managed). This results that when the members of the pool sync with the local repository on the coordinator, the metadata will not be checked, but 'gpgcheck=1' means the RPM packages are to be checked.